livedata/src/main/java/io/getstream/chat/android/livedata/utils/MessageListItemLiveData.kt

```diff
@@ -0,0 +1,223 @@
+ * - Makes the MessageListItem immutable to prevent future bugs
+ *
+ *
+ * @param currentUser the user who is currently authenticated
+ * @param messageLd a livedata object with the messages
+ * @param readsLd a livedata object with the read state per user
+ * @param typingLd a livedata object with the users who are currently typing
+ * @param isThread if we are in a thread or not. if in a thread we add a threadSeperator in position 1 of the item list
+ * @param dateSeparator function to compare previous and current message and return if we should insert a date separator
+ *
+ * Here's an example:
+ *
+ * MessageListItemLiveData(currentUser, messagesLd, readsLd, typingLd, false) { previous, message ->
+ *   return if (previous==null) {
+ *       true
+ *   } else {
+ *       (message.getCreatedAtOrThrow().time - previous.getCreatedAtOrThrow().time) > (60 * 60 * 3)
+ *   }
+ * }
+ *
+ */
+class MessageListItemLiveData(
+    val currentUser: User,
+    val messagesLd: LiveData<List<Message>>,
+    val readsLd: LiveData<List<ChannelUserRead>>,
+    val typingLd: LiveData<List<User>>? = null,
+    val isThread: Boolean = false,
+    val dateSeparator: ((previous: Message?, current: Message) -> Boolean)? = null
+) : MediatorLiveData<MessageListItemWrapper>() {
+
+    private var hasNewMessages: Boolean = false
+    private var messageItemsBase = listOf<MessageListItem>()
+    private var messageItemsWithReads = listOf<MessageListItem>()
+    private var typingUsers = listOf<User>()
+    private var typingItems = listOf<MessageListItem>()
+
+    private var lastMessageID = ""
+
+    init {
+        addSource(messagesLd) { value ->
+            messagesChanged(value)
+        }
+        addSource(readsLd) { value ->
+            readsChanged(value)
+        }
+        if (typingLd != null) {
+            addSource(typingLd) { value ->
+                typingChanged(value)
+            }
+        }
+    }
+
+    internal fun messagesChanged(messages: List<Message>): List<MessageListItem> {
+        messageItemsBase = groupMessages()
+        messageItemsWithReads = addReads(messageItemsBase, readsLd.value)
+        val out = messageItemsWithReads + typingItems
+        val wrapped = wrapMessages(out)
+        value = wrapped.copy(hasNewMessages = hasNewMessages)
+        return out
+    }
+
+    internal fun readsChanged(reads: List<ChannelUserRead>): List<MessageListItem> {
+        messageItemsWithReads = addReads(messageItemsBase, readsLd.value)
+        val out = messageItemsWithReads + typingItems
+        value = wrapMessages(out)
+        return out
+    }
+
+    /**
+     * Typing changes are the most common changes on the message list
+     * Note how they don't recompute the message list, but only add to the end
+     */
+    internal fun typingChanged(users: List<User>): List<MessageListItem> {
+        val newTypingUsers = users.filter { it.id != currentUser.id }
+
+        return if (newTypingUsers != typingUsers) {
```

Comment:
You even can avoid storing current typingUsers. You can do the same in such way:
``` kotlin 
internal fun typingChanged(users: List<User>) {
        val newTypingUsers = users.filter { it.id != currentUser.id }
        val typingItems = usersAsTypingItems(newTypingUsers)
        val messageItemsWithReads = addReads(groupMessages(), readsLd.value)
        val newValue = wrapMessages(messageItemsWithReads + typingItems)
        if (value != newValue) {
            value = newValue
        } 
    }
```